File: hooks/subagent-bootstrap.py
#!/usr/bin/env python3
"""subagentStart hook: link subagent runs to checked-in role prompts.

Fires before a subagent (Task tool) spawns. The hook always allows the
subagent to start. When the `subagent_type` matches a checked-in role under
`agents/`, the hook adds a short `user_message` pointing at the
matching prompt file so the subagent uses the repo-owned definition rather
than reinventing one.
"""
from __future__ import annotations

import json
import os
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from _repo import resolve_repo_root  # noqa: E402
from _trace import trace as _trace  # noqa: E402
from _active_role import set_active_role  # noqa: E402

ROOT = resolve_repo_root(__file__)
AGENTS_DIR = ROOT / "agents"

KNOWN_ROLES = {
    "architect",
    "code-reviewer",
    "critic",
    "debugger",
    "explore",
    "implementer",
    "orchestrator",
    "planner",
    "qa-tester",
    "researcher",
    "security-reviewer",
    "test-engineer",
    "tracer",
    "verifier",
}


def _read_payload() -> dict:
    raw = sys.stdin.read()
    if not raw.strip():
        return {}
    try:
        value = json.loads(raw)
    except json.JSONDecodeError:
        return {"_invalid_json": True}
    return value if isinstance(value, dict) else {"payload": value}


def main() -> int:
    payload = _read_payload()
    if payload.get("_invalid_json"):
        print(json.dumps({"status": "pass", "fail_open": True, "permission": "allow", "user_message": "Subagent-bootstrap input was not JSON; skipped."}))
        return 0

    subagent_type = ""
    for key in ("subagent_type", "subagentType", "type"):
        value = payload.get(key)
        if isinstance(value, str) and value:
            subagent_type = value.lower()
            break

    role_match = subagent_type in KNOWN_ROLES
    role_prompt_path = AGENTS_DIR / f"{subagent_type}.md" if role_match else None
    role_prompt_exists = bool(role_prompt_path and role_prompt_path.is_file())

    if role_match and role_prompt_exists:
        subagent_id = ""
        for key in ("subagent_id", "subagentId", "session_id", "sessionId"):
            value = payload.get(key)
            if isinstance(value, str) and value:
                subagent_id = value
                break
        try:
            set_active_role(subagent_type, subagent_id=subagent_id)
        except OSError:
            pass
        user_message = (
            f"Subagent-bootstrap: matched role `{subagent_type}`. "
            f"Use the checked-in prompt at agents/{subagent_type}.md"
        )
        if subagent_type not in ("implementer", "debugger"):
            user_message += " and stay within its readonly contract."
        else:
            user_message += "."
        status = "matched"
    else:
        user_message = (
            "Subagent-bootstrap: no matching repo-owned role prompt for "
            + (f"`{subagent_type}`" if subagent_type else "this subagent")
            + ". Free-form subagent runs are allowed; check agents/ for a closer match."
        )
        status = "pass"

    output = {
        "status": status,
        "fail_open": True,
        "permission": "allow",
        "user_message": user_message,
        "subagent_type": subagent_type,
        "role_prompt": role_prompt_path.relative_to(ROOT).as_posix() if role_prompt_path else None,  # type: ignore[union-attr]
    }
    _trace({
        "hook": "subagent-bootstrap",
        "event": "subagentStart",
        "status": status,
        "subagent_type": subagent_type,
        "role_match": role_match,
        "role_prompt_exists": role_prompt_exists,
    })
    print(json.dumps(output, ensure_ascii=False))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: hooks/subagent-bootstrap.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as process from 'node:process';
import { fileURLToPath } from 'node:url';
import { resolveRepoRoot } from './_repo.ts';
import { trace } from './_trace.ts';
import { setActiveRole } from './_active_role.ts';

const currentFile = fileURLToPath(import.meta.url);
const ROOT = resolveRepoRoot(currentFile);
const AGENTS_DIR = path.join(ROOT, 'agents');
const SUBAGENT_RUNS_PATH = path.join(ROOT, '.cursor', 'state', 'subagent-runs.json');
const SUBAGENT_RUNS_CAP = 20;

export const KNOWN_ROLES = new Set([
  'architect',
  'code-reviewer',
  'critic',
  'debugger',
  'explore',
  'implementer',
  'orchestrator',
  'planner',
  'qa-tester',
  'researcher',
  'security-reviewer',
  'test-engineer',
  'tracer',
  'verifier',
]);

function readPayload(): any {
  try {
    const raw = fs.readFileSync(0, 'utf-8');
    if (!raw.trim()) {
      return {};
    }
    return JSON.parse(raw);
  } catch (err: any) {
    return { _invalid_json: true };
  }
}

function main(): number {
  const payload = readPayload();
  if (payload?._invalid_json) {
    console.log(JSON.stringify({
      status: 'pass',
      fail_open: true,
      permission: 'allow',
      user_message: 'Subagent-bootstrap input was not JSON; skipped.'
    }));
    return 0;
  }

  let subagentType = '';
  for (const key of ['subagent_type', 'subagentType', 'type']) {
    const value = payload?.[key];
    if (typeof value === 'string' && value) {
      subagentType = value.toLowerCase();
      break;
    }
  }

  const roleMatch = KNOWN_ROLES.has(subagentType);
  const rolePromptPath = roleMatch ? path.join(AGENTS_DIR, `${subagentType}.md`) : null;
  const rolePromptExists = !!(rolePromptPath && fs.existsSync(rolePromptPath) && fs.statSync(rolePromptPath).isFile());

  let userMessage = '';
  let status = 'pass';

  let subagentId = '';
  for (const key of ['subagent_id', 'subagentId', 'session_id', 'sessionId']) {
    const value = payload?.[key];
    if (typeof value === 'string' && value) {
      subagentId = value;
      break;
    }
  }

  const startTs = Date.now();

  if (roleMatch && rolePromptExists) {
    try {
      setActiveRole(subagentType, subagentId);
    } catch {
      // ignore
    }
    userMessage = `Subagent-bootstrap: matched role \`${subagentType}\`. Use the checked-in prompt at agents/${subagentType}.md`;
    if (subagentType !== 'implementer' && subagentType !== 'debugger') {
      userMessage += ' and stay within its readonly contract.';
    } else {
      userMessage += '.';
    }
    status = 'matched';
  } else {
    userMessage = (
      'Subagent-bootstrap: no matching repo-owned role prompt for ' +
      (subagentType ? `\`${subagentType}\`` : 'this subagent') +
      '. Free-form subagent runs are allowed; check agents/ for a closer match.'
    );
    status = 'pass';
  }

  const relativePromptPath = rolePromptPath ? path.relative(ROOT, rolePromptPath) : null;
  const output = {
    status,
    fail_open: true,
    permission: 'allow',
    user_message: userMessage,
    subagent_type: subagentType,
    role_prompt: relativePromptPath,
  };

  try {
    fs.mkdirSync(path.dirname(SUBAGENT_RUNS_PATH), { recursive: true });
    let runs: any[] = [];
    if (fs.existsSync(SUBAGENT_RUNS_PATH)) {
      try {
        const parsed = JSON.parse(fs.readFileSync(SUBAGENT_RUNS_PATH, 'utf-8'));
        if (Array.isArray(parsed)) runs = parsed;
      } catch {
        runs = [];
      }
    }
    runs.push({ role: subagentType, start_ts: startTs, subagent_id: subagentId || subagentType });
    if (runs.length > SUBAGENT_RUNS_CAP) {
      runs = runs.slice(runs.length - SUBAGENT_RUNS_CAP);
    }
    fs.writeFileSync(SUBAGENT_RUNS_PATH, JSON.stringify(runs, null, 2));
  } catch {
    // fail-open
  }

  trace({
    hook: 'subagent-bootstrap',
    event: 'subagentStart',
    status,
    subagent_type: subagentType,
    role_match: roleMatch,
    role_prompt_exists: rolePromptExists,
    subagent_start_ts: startTs,
  });

  console.log(JSON.stringify(output));
  return 0;
}

process.exit(main());
